File: Tutorial/go/algorithmPrj/algorithmbygo/tree/tree.go
// @File    : tree
// @Project : go
// @Time    : 2021/5/13 13:44
// ======================================================
//                                        /
//   __.  , , , _  _   __ ______  _    __/  __ ____  _,
//   (_/|_(_(_/_</_/_)_(_)/ / / <_</_  (_/_ (_)/ / <_(_)_
//                                                    /|
//                                                   |/
// ======================================================
package tree

type Tree struct {
	Val   int
	left  *Tree
	right *Tree
}

func GenerateTreeByArray(nums []int) *Tree {
	if len(nums) == 0 {
		return nil
	}
	if len(nums) == 1 {
		return &Tree{nums[0], nil, nil}
	}
	root := &Tree{nums[0], nil, nil}
	nodes := []*Tree{root}
	nums = append(nums[1:])
	for len(nums) != 0 {
		for _, node := range nodes {
			nodes = append(nodes[1:])
			if len(nums) != 0 {
				node.left = &Tree{nums[0], nil, nil}
				nums = append(nums[1:])
				nodes = append(nodes, node.left)
			}
			if len(nums) != 0 {
				node.right = &Tree{nums[0], nil, nil}
				nums = append(nums[1:])
				nodes = append(nodes, node.right)
			}

		}
	}
	return root
}

func helper(tree *Tree, res *[]int) {
	if tree != nil {
		helper(tree.left, res)
		*res = append(*res, tree.Val)
		helper(tree.right, res)
	}
}

func IterateTreeDFSInorder(tree *Tree) []int {
	var ans []int

	helper(tree, &ans)

	return ans
}

File: Interview/tree.py
# -*- coding: utf-8 -*-
# @File    : tree
# @Project : 4U
# @Time    : 2024/5/29 14:42
# once, I want 2 leave my name 2 the history
# sadly, I find that my hair gets pale
"""                     
                                      /             
 __.  , , , _  _   __ ______  _    __/  __ ____  _, 
(_/|_(_(_/_</_/_)_(_)/ / / <_</_  (_/_ (_)/ / <_(_)_
                                                 /| 
                                                |/  
"""
from Interview.func import construct_binary_tree, gen_random_list


def dfs(root):
    if root is not None:
        yield from dfs(root.left)
        yield root.value
        yield from dfs(root.right)


def bfs(root):
    if root is not None:
        queue = [root]
        while queue:
            node = queue.pop(0)
            yield node.value
            queue.extend([i for i in (node.left, node.right) if i is not None])


if __name__ == '__main__':
    tree = construct_binary_tree(gen_random_list())
    print(list(dfs(tree)))
    print(list(bfs(tree)))
